### PR TITLE
Coverage report

### PR DESCRIPTION
**Pull request by wzalazar**
### Description of Changes

Introduces report of coverage for unit test and integration test.

I run the coverage separated because are different instances and when you run another after the first one it will rewrite the input. Maybe when we have the all test in the same instance we will be able to capture the all the coverage together.

needed-by https://github.com/poetapp/node/issues/11

### PR Review Checklist

* [ ] **Have you tested the changes?**
* [ ] **Is QE review required? If so, has this been reviewed by QE?** (For all new features, review by QE is required. Minor changes which already have passing functional tests may not need a new QE review -- when in doubt, loop in a QE reviewer).
* [ ] **Are there code changes that require new unit tests? If so, are those tests included?** (A separate PR for tests is not appropriate).
* [ ] **Are documentation updates required? If so, does this PR include it?** (Separate documentation PR is generally not appropriate).
* [ ] **Are the added/edited files named appropriately and located correctly in the project directory structure?**

### CODE REVIEW

**Comment by wzalazar** [comment]
the scope is only for showing the coverage report.

**Comment by lautarodragan** [comment]
Thanks.

I'm seeing the following output in Travis:

For unit tests (RITEway):
```
=============================== Coverage summary ===============================
Statements   : 13.33% ( 12/90 )
Branches     : 23.33% ( 7/30 )
Functions    : 10% ( 2/20 )
Lines        : 13.41% ( 11/82 )
================================================================================
```

For integration tests (Alsatian):
```
=============================== Coverage summary ===============================
Statements   : Unknown% ( 0/0 )
Branches     : Unknown% ( 0/0 )
Functions    : Unknown% ( 0/0 )
Lines        : Unknown% ( 0/0 )
================================================================================
```

Is this the expected output? What's up with the `Unknown% ( 0/0 )`?

**Comment by lautarodragan** [comment]
Also: it's a good opportunity to update the README's coverage section.

https://github.com/poetapp/node/blob/6a00eaff4ae8145fba9efff3d0ab8e5a271c8f13/README.md#L277-L284

**Comment by wzalazar** [comment]
@lautarodragan 
I will review the report of coverage of the integration test, but I'm thinking is because I added all files.

**Comment by wzalazar** [comment]
Regarding integration test:

well, before shown wrong information because it was over test files
```
-------------------------|----------|----------|----------|----------|-------------------|
File                     |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
-------------------------|----------|----------|----------|----------|-------------------|
All files                |    97.87 |      100 |       80 |    97.81 |                   |
 test                    |    66.67 |      100 |        0 |    66.67 |                   |
  Claims.ts              |     62.5 |      100 |        0 |     62.5 |          64,65,66 |
  Keys.ts                |      100 |      100 |      100 |      100 |                   |
 test/Integration        |      100 |      100 |      100 |      100 |                   |
  GetWork.ts             |      100 |      100 |      100 |      100 |                   |
  GetWorksByPublicKey.ts |      100 |      100 |      100 |      100 |                   |
  Helper.ts              |      100 |      100 |      100 |      100 |                   |
  PostWork.ts            |      100 |      100 |      100 |      100 |                   |
  SecurityHeaders.ts     |      100 |      100 |      100 |      100 |                   |
-------------------------|----------|----------|----------|----------|-------------------|
```

now the information it's about the app files

```
------------------------------|----------|----------|----------|----------|-------------------|
File                          |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
------------------------------|----------|----------|----------|----------|-------------------|
All files                     |        0 |        0 |        0 |        0 |                   |
 src                          |        0 |        0 |        0 |        0 |                   |
  Error.ts                    |        0 |        0 |      100 |        0 |             1,2,4 |
  index.ts                    |        0 |        0 |        0 |        0 |... 09,110,112,116 |
 src/API                      |        0 |      100 |      100 |        0 |                   |
  SecurityHeaders.ts          |        0 |      100 |      100 |        0 |                 1 |
 src/API/Middlewares          |        0 |        0 |        0 |        0 |                   |
  HttpExceptionsMiddleware.ts |        0 |        0 |        0 |        0 |... 14,16,17,19,20 |
  LoggerMiddleware.ts         |        0 |      100 |        0 |        0 |             4,8,9 |
 src/Extensions               |        0 |        0 |        0 |        0 |                   |
  Promise.ts                  |        0 |        0 |        0 |        0 |             5,8,9 |
 src/Helpers                  |        0 |        0 |        0 |        0 |                   |
  AsyncPipe.ts                |        0 |      100 |        0 |        0 |                 3 |
  Bitcoin.ts                  |        0 |        0 |        0 |        0 |... 11,13,25,29,30 |
  Logging.ts                  |        0 |      100 |        0 |        0 |        7,11,15,18 |
  Time.ts                     |        0 |      100 |        0 |        0 |               1,2 |
------------------------------|----------|----------|----------|----------|-------------------|
=============================== Coverage summary ===============================
Statements   : 0% ( 0/77 )
Branches     : 0% ( 0/22 )
Functions    : 0% ( 0/18 )
Lines        : 0% ( 0/70 )
================================================================================
```

**Comment by wzalazar** [comment]
well, we have a problem with the option `--all` of `nyc`. This option is omitting several files. I'm researching about it, I found a lot of solutions but none was fixed the problem.

**Comment by lautarodragan** [comment]
If it's only a problem with Alsatian maybe we can ignore it?

**Comment by wzalazar** [comment]
Report:

**Unit test**: it's already. Several files are untracked until we will the unit test.

**Integrarion test**: At the moment there are several issues with this report. there is no official documentation about how to do get the report for the integration tests.

I left this issue, is like an introduction to the problem
https://github.com/istanbuljs/nyc/issues/548

@lautarodragan @warrenv 

Keeping in mind that the report doesn't work properly. What is your opinion with the report for the integration test?


When we decided the path, I will update the README

**Comment by warrenv** [comment]
Is it possible report on untested files for the unit test? Or do they also have a problem with the --all flag?

**Comment by wzalazar** [comment]
@warrenv 

> Is it possible report on untested files for the unit test? Or do they also have a problem with the --all flag?

the `--all` option is in `true`. But all files that contain 

- interface
- decorator 

these will be untrucked when you do the unit test.

Is a bug, I think yes

**Comment by warrenv** [comment]
ok. so typescript is interfering here somehow?
one of the purposes of running coverage reports is to expose files that don't have unit tests or are missing sufficient coverage.

**Comment by wzalazar** [comment]
seems is Typescript, I leave these links

**Coverage reports don't include TS files unless explicitly compiled** 
https://github.com/istanbuljs/nyc/issues/742
open: Dec 8 2017
last comment: Jun 15 2018 
status: OPEN

**`--all` does not include `.ts` files that use typecasting**
https://github.com/istanbuljs/nyc/issues/679
open: Sep 22 2017
last comment: Jun 16 2018
status: OPEN


**Typescript code coverage with "all": true flag reports incorrect missing coverage**
https://github.com/istanbuljs/nyc/pull/643
open: Aug 11 2017
last comment: May 17 2018
status: CLOSED

**Problems with --all**
https://github.com/istanbuljs/nyc/issues/603 (Jun 11 2017 - Mar 29 2018) OPEN
open: Jun 11 2017
last comment: Mar 29 2018
status: OPEN


**--all fails with ts-node (Typescript)**
https://github.com/istanbuljs/nyc/issues/504
open: Jan 26 2017
last comment: 1 Jul 2018
status: CLOSED


**Including all files when not using require or register**
https://github.com/istanbuljs/nyc/issues/776
open: Feb 10 2018
last comment: Feb 17 2018
status: OPEN


**Better TypeScript coverage support in Istanbul**
https://github.com/Microsoft/TypeScript/issues/24993
open: Jun 15 2018
last comment: Jun 15 2018
status: OPEN

**Comment by warrenv** [comment]
thanks for thoroughly documenting that Walter!

If we made sure there was a unit test file to every code file (even if it
only ran the most basic test),
would that get us around the problem? Every file would be required so every
file by definition would
show up in the coverage report.

We could write a script that makes sure every file has a matching unit test
file and have travis run it so newly added files don't sneak thru.

**Comment by wzalazar** [comment]
well, I found the real reason. If you run this command you can see the log errors:

`NODE_DEBUG=nyc npm run coverage:unit`

so we get this results in several times:

> NYC 10267: failed to instrument /src/View/WorkController.tswith error: SyntaxError: This experimental syntax requires enabling one of the following parser plugin(s): 'decorators-legacy, decorators' (9:0)

Because the files always are instrumented by default:

 ```
    try {
      instrumented = instrumenter.instrumentSync(code, filename, sourceMap)
    } catch (e) {
      debugLog('failed to instrument ' + filename + ' with error: ' + e.stack)
      if (this.config.exitOnError) {
        console.error('Failed to instrument ' + filename)
        process.exit(1)
      } else {
        instrumented = code
      }
    }


    if (this.fakeRequire) {
      return 'function x () {}'
    } else {
      return instrumented
    }
  }
}
```
https://github.com/istanbuljs/nyc/blob/6b6cd5e5cb9469aa1975e41d9950783832d7b815/index.js#L281-L299

you have the option "instrument": false, but if you set in `false`, the files without a explicit `require` won't appear.

`nyc` it is using [this package](https://github.com/istanbuljs/istanbuljs/tree/master/packages/istanbul-lib-instrument) to instrument

if you can see in this lines,

```
  const ast = parser.parse(code, {
            allowReturnOutsideFunction: opts.autoWrap,
            sourceType: opts.esModules ? "module" : "script",
            plugins: [
              'asyncGenerators',
              'dynamicImport',
              'objectRestSpread',
              'optionalCatchBinding',
              'flow',
              'jsx'
            ]
```
https://github.com/istanbuljs/istanbuljs/blob/d23d023d43c7b030e708518e6791b67f39c4c9ad/packages/istanbul-lib-instrument/src/instrumenter.js#L83-L93

the plugin for supporting decorators doesn't appear and maybe it is not the only problem

**Comment by wzalazar** [comment]
well, I played around istanbul-lib-instrument adding the plugins and surprise, a lot of issues was resolved

from (original):
```
plugins: ['asyncGenerators', 'dynamicImport', 'objectRestSpread', 'optionalCatchBinding', 'flow', 'jsx']
```

to:

```
plugins: ['asyncGenerators', 'dynamicImport', 'objectRestSpread', 'optionalCatchBinding', 'jsx', 'typescript', 'classProperties', ['decorators', { decoratorsBeforeExport: true }]]
```

but, we are using parameter decorators in our base code so we receive this issue

> NYC 13853: failed to instrument /src/View/WorkController.tswith error: SyntaxError: Stage 2 decorators cannot be used to decorate parameters (15:14)

I reserched about it:

https://github.com/babel/babel/issues/1301#issuecomment-94337907

https://github.com/rstuven/es-decorate/issues/1 
Apr 28, 2015
status: OPEN

However the report included more files:

```
------------------------------------------|----------|----------|----------|----------|-------------------|
File                                      |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
------------------------------------------|----------|----------|----------|----------|-------------------|
All files                                 |     4.26 |    10.94 |     2.78 |     4.15 |                   |
 src                                      |        0 |        0 |        0 |        0 |                   |
  Configuration.ts                        |        0 |        0 |        0 |        0 |... 06,107,108,116 |
  Error.ts                                |        0 |        0 |      100 |        0 |             1,2,4 |
  Interfaces.ts                           |        0 |        0 |        0 |        0 |                22 |
  index.ts                                |        0 |        0 |        0 |        0 |... 09,110,112,116 |
 src/API                                  |        0 |      100 |        0 |        0 |                   |
  API.ts                                  |        0 |      100 |        0 |        0 |... 45,46,47,48,49 |
  APIConfiguration.ts                     |        0 |        0 |        0 |        0 |                   |
  Grettings.ts                            |        0 |      100 |        0 |        0 |               7,8 |
  RouterConfiguration.ts                  |        0 |        0 |        0 |        0 |                   |
  SecurityHeaders.ts                      |        0 |      100 |      100 |        0 |                 1 |
 src/API/Middlewares                      |       40 |       50 |       40 |    39.29 |                   |
  HttpExceptionsMiddleware.ts             |        0 |        0 |        0 |        0 |... 14,16,17,19,20 |
  LoggerMiddleware.ts                     |        0 |      100 |        0 |        0 |             4,8,9 |
  RequestValidationMiddleware.ts          |    92.31 |     87.5 |      100 |    91.67 |                24 |
 src/BlockchainReader                     |        0 |      100 |        0 |        0 |                   |
  BlockchainReader.ts                     |        0 |      100 |        0 |        0 |... 49,50,51,54,57 |
  BlockchainReaderConfiguration.ts        |        0 |        0 |        0 |        0 |                   |
  BlockchainReaderServiceConfiguration.ts |        0 |        0 |        0 |        0 |                   |
  ClaimControllerConfiguration.ts         |        0 |        0 |        0 |        0 |                   |
 src/BlockchainWriter                     |        0 |      100 |        0 |        0 |                   |
  BlockchainWriter.ts                     |        0 |      100 |        0 |        0 |... 61,64,65,71,72 |
  BlockchainWriterConfiguration.ts        |        0 |        0 |        0 |        0 |                   |
  ClaimControllerConfiguration.ts         |        0 |        0 |        0 |        0 |                   |
  ServiceConfiguration.ts                 |        0 |        0 |        0 |        0 |                   |
 src/Extensions                           |        0 |        0 |        0 |        0 |                   |
  Promise.ts                              |        0 |        0 |        0 |        0 |             5,8,9 |
 src/Helpers                              |        0 |        0 |        0 |        0 |                   |
  AsyncPipe.ts                            |        0 |      100 |        0 |        0 |                 3 |
  Bitcoin.ts                              |        0 |        0 |        0 |        0 |... 11,13,25,29,30 |
  FetchError.ts                           |        0 |        0 |        0 |        0 |          15,17,19 |
  Logging.ts                              |        0 |      100 |        0 |        0 |        7,11,15,18 |
  Time.ts                                 |        0 |      100 |        0 |        0 |               1,2 |
 src/Messaging                            |        0 |        0 |        0 |        0 |                   |
  Messages.ts                             |        0 |        0 |        0 |        0 |                   |
  Messaging.ts                            |        0 |        0 |        0 |        0 |... 03,106,108,111 |
 src/Storage                              |        0 |        0 |        0 |        0 |                   |
  ClaimControllerConfiguration.ts         |        0 |        0 |        0 |        0 |                   |
  DownloadFailure.ts                      |        0 |        0 |        0 |        0 |                   |
  Entry.ts                                |        0 |        0 |        0 |        0 |                   |
  Exceptions.ts                           |        0 |        0 |        0 |        0 |... 30,38,39,43,44 |
  IPFSConfiguration.ts                    |        0 |        0 |        0 |        0 |                   |
  ServiceConfiguration.ts                 |        0 |        0 |        0 |        0 |                   |
  Storage.ts                              |        0 |      100 |        0 |        0 |... 68,69,75,76,77 |
  StorageConfiguration.ts                 |        0 |        0 |        0 |        0 |                   |
 src/View                                 |        0 |      100 |        0 |        0 |                   |
  View.ts                                 |        0 |      100 |        0 |        0 |... 43,44,45,46,47 |
  ViewConfiguration.ts                    |        0 |        0 |        0 |        0 |                   |
------------------------------------------|----------|----------|----------|----------|-------------------|

=============================== Coverage summary ===============================
Statements   : 4.26% ( 12/282 )
Branches     : 10.94% ( 7/64 )
Functions    : 2.78% ( 2/72 )
Lines        : 4.15% ( 11/265 )
================================================================================
```

in this time was omitted the files that implement a decorator in the parameters

**Comment by lautarodragan** [comment]
Amazing work @wzalazar!

**Comment by wzalazar** [comment]
I created these PRs in `istanbul-lib-instrument`

https://github.com/istanbuljs/istanbuljs/pull/204
https://github.com/istanbuljs/istanbuljs/pull/205

if these PRs are merged after that we need to create a PR in `nyc` for that accepting the new options for the instrument

**Comment by wzalazar** [comment]
our `nyc` is working now, this is the output the coverage of unit test:

```
------------------------------------------|----------|----------|----------|----------|-------------------|
File                                      |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
------------------------------------------|----------|----------|----------|----------|-------------------|
All files                                 |     0.74 |     3.45 |        0 |     0.79 |                   |
 src                                      |        0 |        0 |        0 |        0 |                   |
  Configuration.ts                        |        0 |        0 |        0 |        0 |... 06,107,108,116 |
  Error.ts                                |        0 |        0 |        0 |        0 |             1,2,4 |
  Interfaces.ts                           |        0 |        0 |        0 |        0 |                22 |
  index.ts                                |        0 |        0 |        0 |        0 |... 09,110,112,116 |
 src/API                                  |        0 |      100 |        0 |        0 |                   |
  API.ts                                  |        0 |      100 |        0 |        0 |... 45,46,47,48,49 |
  APIConfiguration.ts                     |        0 |        0 |        0 |        0 |                   |
  RouterConfiguration.ts                  |        0 |        0 |        0 |        0 |                   |
  SecurityHeaders.ts                      |        0 |      100 |      100 |        0 |                 1 |
 src/API/Middlewares                      |    10.53 |       25 |        0 |    11.11 |                   |
  HttpExceptionsMiddleware.ts             |        0 |        0 |        0 |        0 |... 14,16,17,19,20 |
  LoggerMiddleware.ts                     |        0 |      100 |        0 |        0 |             4,8,9 |
  RequestValidationMiddleware.ts          |      100 |      100 |      100 |      100 |                   |
 src/BlockchainReader                     |        0 |      100 |        0 |        0 |                   |
  BlockchainReader.ts                     |        0 |      100 |        0 |        0 |... 49,50,51,54,57 |
  BlockchainReaderConfiguration.ts        |        0 |        0 |        0 |        0 |                   |
  BlockchainReaderServiceConfiguration.ts |        0 |        0 |        0 |        0 |                   |
  ClaimControllerConfiguration.ts         |        0 |        0 |        0 |        0 |                   |
 src/BlockchainWriter                     |        0 |      100 |        0 |        0 |                   |
  BlockchainWriter.ts                     |        0 |      100 |        0 |        0 |... 61,64,65,71,72 |
  BlockchainWriterConfiguration.ts        |        0 |        0 |        0 |        0 |                   |
  ClaimControllerConfiguration.ts         |        0 |        0 |        0 |        0 |                   |
  ServiceConfiguration.ts                 |        0 |        0 |        0 |        0 |                   |
 src/Extensions                           |        0 |        0 |        0 |        0 |                   |
  Promise.ts                              |        0 |        0 |        0 |        0 |             5,8,9 |
 src/Helpers                              |        0 |        0 |        0 |        0 |                   |
  AsyncPipe.ts                            |        0 |      100 |        0 |        0 |                 3 |
  Bitcoin.ts                              |        0 |        0 |        0 |        0 |... 11,13,25,29,30 |
  FetchError.ts                           |        0 |        0 |        0 |        0 |          15,17,19 |
  Logging.ts                              |        0 |      100 |        0 |        0 |        7,11,15,18 |
  Time.ts                                 |        0 |      100 |        0 |        0 |               1,2 |
 src/Messaging                            |        0 |        0 |        0 |        0 |                   |
  Messages.ts                             |        0 |        0 |        0 |        0 |                   |
  Messaging.ts                            |        0 |        0 |        0 |        0 |... 03,106,108,111 |
 src/Storage                              |        0 |        0 |        0 |        0 |                   |
  ClaimControllerConfiguration.ts         |        0 |        0 |        0 |        0 |                   |
  DownloadFailure.ts                      |        0 |        0 |        0 |        0 |                   |
  Entry.ts                                |        0 |        0 |        0 |        0 |                   |
  Exceptions.ts                           |        0 |        0 |        0 |        0 |... 30,38,39,43,44 |
  IPFSConfiguration.ts                    |        0 |        0 |        0 |        0 |                   |
  ServiceConfiguration.ts                 |        0 |        0 |        0 |        0 |                   |
  Storage.ts                              |        0 |      100 |        0 |        0 |... 68,69,75,76,77 |
  StorageConfiguration.ts                 |        0 |        0 |        0 |        0 |                   |
 src/View                                 |        0 |      100 |        0 |        0 |                   |
  View.ts                                 |        0 |      100 |        0 |        0 |... 43,44,45,46,47 |
  ViewConfiguration.ts                    |        0 |        0 |        0 |        0 |                   |
------------------------------------------|----------|----------|----------|----------|-------------------|
=============================== Coverage summary ===============================
Statements   : 0.74% ( 2/269 )
Branches     : 3.45% ( 2/58 )
Functions    : 0% ( 0/70 )
Lines        : 0.79% ( 2/253 )
================================================================================
```

File: package.json [hunk 1]
```diff
@@ -70,7 +70,7 @@
     "babel-plugin-module-resolver": "3.1.1",
     "concurrently": "3.6.0",
     "nodemon": "1.18.3",
-    "nyc": "12.0.2",
+    "nyc": "git+https://github.com/poetapp/nyc.git",
```
Comment (lautarodragan) [inline package.json]: Can you specify a commit hash? This way if changes are pushed the master branch of poetapp/nic they won't get picked up automatically by the node.

File: package.json [hunk 2]
```diff
@@ -68,10 +70,11 @@
     "babel-plugin-module-resolver": "3.1.1",
     "concurrently": "3.6.0",
     "nodemon": "1.18.3",
-    "nyc": "12.0.2",
+    "nyc": "https://github.com/poetapp/nyc.git",
```
Comment (lautarodragan) [inline package.json]: Can you specify a commit hash? This way if changes are pushed the master branch of poetapp/nic they won't get picked up automatically by the node.

**Comment by lautarodragan** [comment]
Looks like there's still some files missing from that report.

In src/API: Router.ts, WorkController.ts

File: package.json (hunk 2 above)
Comment (wzalazar) [reply package.json]: yes, good idea!!!  :+1:

**Comment by wzalazar** [comment]
> Looks like there's still some files missing from that report.
> 
> In src/API: Router.ts, WorkController.ts

these files have a problem with the instrument because it has a decorator in the constructor. I have to create a unit test for this files.

**Comment by wzalazar** [comment]
@warrenv thanks!!!

from here we start...

```
=============================== Coverage summary ===============================
Statements   : 34.85% ( 237/680 )
Branches     : 4.35% ( 4/92 )
Functions    : 11.76% ( 14/119 )
Lines        : 33.87% ( 210/620 )
================================================================================
```

**Comment by wzalazar** [comment]
@lautarodragan @warrenv 

I need your review

**Pull request merged**